### PR TITLE
HPCC-28723 Fix potential PARALLEL crash

### PR DESCRIPTION
**Pull request by jakesmith**
A PARALLEL activity might not have all input branches connected. Validate inputs before traversing.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
@jakesmith one question - I'm not sure why input would be null

File: thorlcr/graph/thgraph.cpp [hunk 1]
```diff
@@ -763,16 +763,18 @@ bool CGraphElementBase::prepareContext(size32_t parentExtractSz, const byte *par
...
+                    if (!whichInput->prepareContext(parentExtractSz, parentExtract, true, false, async, connectOnly))
                         return false;
                 }
                 ForEachItemIn(i, inputs)
                 {
                     if (i != whichBranch)
                     {
-                        if (!queryInput(i)->prepareContext(parentExtractSz, parentExtract, false, false, async, true))
+                        CGraphElementBase *input = queryInput(i);
+                        if (!input->prepareContext(parentExtractSz, parentExtract, false, false, async, true))
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply thorlcr/graph/thgraph.cpp:777]: yes, I meant there to be

[the remark this answers is not in the document]
Comment (jakesmith) [reply thorlcr/graph/thgraph.cpp]: @ghalliday - squashed change in.

**Comment by jakesmith** [comment]
> I'm not sure why input would be null

maybe PARALLEL (and SEQUENTIAL) are special cases, where they can have multiple inputs, but where only some are connected ?
But I changed so that it guarded against that possibility in general .. 

I attached an example of the PARALLEL case that reproduced the problem to the JIRA.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**